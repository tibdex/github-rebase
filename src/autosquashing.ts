import {
  CommitDetails,
  CommitMessage,
  Sha,
} from "shared-github-internals/lib/git";

type AutosquashingMode = null | "fixup" | "squash";

type AutosquashingStep = {
  autosquashMessage: null | CommitMessage;
  shas: Sha[];
};

const getCommitSubjectAndBody = (commitMessage: CommitMessage) => {
  const [subject, ...rest] = commitMessage.split(/(\r\n|\r|\n){2}/u);
  return {
    body: rest
      .map(line => line.trim())
      .filter(line => line !== "")
      .join("\n"),
    subject,
  };
};

const getAutosquashMode = ({
  commitDetails,
  message,
}: {
  commitDetails: CommitDetails;
  message: CommitMessage;
}) => {
  // It's fine, the data is coming from the GitHub API,
  // it won't have a weird shape.
  // eslint-disable-next-line security/detect-non-literal-regexp
  const matches = new RegExp(
    `^(fixup|squash)! (fixup! |squash! )*(${
      getCommitSubjectAndBody(commitDetails.message).subject
    }|${commitDetails.sha}|${commitDetails.sha.substr(7)})$`,
    "u",
  ).exec(getCommitSubjectAndBody(message).subject);
  if (!matches) {
    return null;
  }
  return matches[1] === "fixup" ? "fixup" : "squash";
};

const getNewAutosquashMessage = ({
  commitsDetails,
  message,
  mode,
  step,
}: {
  commitsDetails: CommitDetails[];
  message: CommitMessage;
  mode: AutosquashingMode;
  step: AutosquashingStep;
}) => {
  const previousMessage =
    step.autosquashMessage === null
      ? // We know that the commit details will be found.
        // @ts-ignore
        commitsDetails.find(({ sha }) => sha === step.shas[0]).message
      : step.autosquashMessage;
  return mode === "squash"
    ? `${previousMessage}\n\n${message}`
    : previousMessage;
};

const groupNonAutosquashingSteps = ({
  newStep,
  steps,
}: {
  newStep: AutosquashingStep;
  steps: AutosquashingStep[];
}) =>
  newStep.autosquashMessage === null &&
  steps.length > 0 &&
  steps[steps.length - 1].autosquashMessage === null
    ? [
        ...steps.slice(0, -1),
        {
          autosquashMessage: null,
          shas: [...steps[steps.length - 1].shas, ...newStep.shas],
        },
      ]
    : [...steps, newStep];

const getAutosquashingSteps = (commitsDetails: CommitDetails[]) => {
  const alreadyHandledShas = new Set();
  const initialSteps: AutosquashingStep[] = [];

  return commitsDetails.reduce((steps, commitDetails) => {
    if (alreadyHandledShas.has(commitDetails.sha)) {
      return steps;
    }

    alreadyHandledShas.add(commitDetails.sha);

    const initialStep: AutosquashingStep = {
      autosquashMessage: null,
      shas: [commitDetails.sha],
    };

    const newStep = commitsDetails
      .filter(({ sha }) => !alreadyHandledShas.has(sha))
      .reduce((step, { message, sha }) => {
        const mode = getAutosquashMode({ commitDetails, message });
        if (mode === null) {
          return step;
        }
        alreadyHandledShas.add(sha);
        return {
          autosquashMessage: getNewAutosquashMessage({
            commitsDetails,
            message,
            mode,
            step,
          }),
          shas: [...step.shas, sha],
        };
      }, initialStep);

    return groupNonAutosquashingSteps({ newStep, steps });
  }, initialSteps);
};

export { AutosquashingStep };

export default getAutosquashingSteps;
