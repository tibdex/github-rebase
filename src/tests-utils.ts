import { Ref } from "shared-github-internals/lib/git";
import {
  CommandDirectory,
  createGitRepo,
  executeGitCommand,
  RepoState,
} from "shared-github-internals/lib/tests/git";

const createGitRepoAndRebase = async ({
  initialState,
  ref,
}: {
  initialState: RepoState;
  ref: Ref;
}): Promise<CommandDirectory> => {
  const directory = await createGitRepo(initialState);
  await executeGitCommand({
    args: ["rebase", "--autosquash", "--interactive", "master"],
    directory,
    // See https://stackoverflow.com/a/29094904
    env: { GIT_EDITOR: ":", GIT_SEQUENCE_EDITOR: ":" },
    ref,
  });
  return directory;
};

export { createGitRepoAndRebase };
