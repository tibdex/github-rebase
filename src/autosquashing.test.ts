import { flatten } from "lodash";
import {
  CommitDetails,
  CommitMessage,
  Ref,
} from "shared-github-internals/lib/git";
import { getRefCommitsFromGitRepo } from "shared-github-internals/lib/tests/git";

import { AutosquashingStep, getAutosquashingSteps } from "./autosquashing";
import { createGitRepoAndRebase } from "./tests-utils";

const getCommit = ({
  commitCounts,
  message,
  position,
}: {
  commitCounts: number;
  message: CommitMessage;
  position: number;
}) => ({
  lines: new Array(commitCounts - 1)
    .fill("initial")
    .map((value, index) => (index < position ? String(index) : value)),
  message,
});

const commitsDetailsToInitialState = ({
  commitsDetails,
  ref,
}: {
  commitsDetails: CommitDetails[];
  ref: Ref;
}) => {
  const commitCounts = commitsDetails.length + 1;
  return {
    initialCommit: getCommit({ commitCounts, message: "initial", position: 0 }),
    refsCommits: {
      [ref]: commitsDetails.map(({ message }, index) =>
        getCommit({ commitCounts, message, position: index + 1 }),
      ),
    },
  };
};

const autosquashingStepsToCommitMessages = ({
  commitsDetails,
  steps,
}: {
  commitsDetails: CommitDetails[];
  steps: AutosquashingStep[];
}) =>
  flatten(
    steps.map(({ autosquashMessage, shas }) =>
      autosquashMessage === null
        ? shas.map(
            stepSha =>
              // @ts-ignore We know that the commit details will be found.
              commitsDetails.find(({ sha }) => sha === stepSha).message,
          )
        : autosquashMessage,
    ),
  );

test.each([
  ["nothing to do with", ["a", "fixup! b"]],
  ["simple fixup", ["a", "b", "fixup! a"]],
  ["fixup first commit only", ["a", "fixup! a", "fixup! a"]],
  [
    "simple squash",
    ["a\n\nSome details", "b", "squash! a\n\nSome more details"],
  ],
  [
    "lot of things to do",
    [
      "a\n\nSome details\non two lines",
      "b",
      "fixup! a\n\nSome unnecessary details",
      "c",
      "d",
      "squash! fixup! a\n\nAgain some more details",
      "squash! b",
      "squash! a\n\nLast fix",
    ],
  ],
])("%s", async (tmp, commitMessages) => {
  const ref = "feature";
  const commitsDetails = commitMessages.map(
    (message: CommitMessage, index: number) => ({
      message,
      sha: new Array(7).fill(index).join(""),
    }),
  );
  const directory = await createGitRepoAndRebase({
    initialState: commitsDetailsToInitialState({
      commitsDetails,
      ref,
    }),
    ref,
  });
  const expectedCommits = await getRefCommitsFromGitRepo({
    directory,
    ref,
  });
  const expectedMessages = expectedCommits
    .slice(1)
    .map(({ message }) => message);
  expect({ commitsDetails, expectedMessages }).toMatchSnapshot();
  const steps = getAutosquashingSteps(commitsDetails);
  expect(autosquashingStepsToCommitMessages({ commitsDetails, steps })).toEqual(
    expectedMessages,
  );
});
