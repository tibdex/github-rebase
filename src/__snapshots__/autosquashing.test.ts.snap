// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`fixup first commit only 1`] = `
Object {
  "commitsDetails": Array [
    Object {
      "message": "a",
      "sha": "0000000",
    },
    Object {
      "message": "fixup! a",
      "sha": "1111111",
    },
    Object {
      "message": "fixup! a",
      "sha": "2222222",
    },
  ],
  "expectedMessages": Array [
    "a",
  ],
}
`;

exports[`lot of things to do 1`] = `
Object {
  "commitsDetails": Array [
    Object {
      "message": "a

Some details
on two lines",
      "sha": "0000000",
    },
    Object {
      "message": "b",
      "sha": "1111111",
    },
    Object {
      "message": "fixup! a

Some unnecessary details",
      "sha": "2222222",
    },
    Object {
      "message": "c",
      "sha": "3333333",
    },
    Object {
      "message": "d",
      "sha": "4444444",
    },
    Object {
      "message": "squash! fixup! a

Again some more details",
      "sha": "5555555",
    },
    Object {
      "message": "squash! b",
      "sha": "6666666",
    },
    Object {
      "message": "squash! a

Last fix",
      "sha": "7777777",
    },
  ],
  "expectedMessages": Array [
    "a

Some details
on two lines

squash! fixup! a

Again some more details

squash! a

Last fix",
    "b

squash! b",
    "c",
    "d",
  ],
}
`;

exports[`nothing to do with 1`] = `
Object {
  "commitsDetails": Array [
    Object {
      "message": "a",
      "sha": "0000000",
    },
    Object {
      "message": "fixup! b",
      "sha": "1111111",
    },
  ],
  "expectedMessages": Array [
    "a",
    "fixup! b",
  ],
}
`;

exports[`simple fixup 1`] = `
Object {
  "commitsDetails": Array [
    Object {
      "message": "a",
      "sha": "0000000",
    },
    Object {
      "message": "b",
      "sha": "1111111",
    },
    Object {
      "message": "fixup! a",
      "sha": "2222222",
    },
  ],
  "expectedMessages": Array [
    "a",
    "b",
  ],
}
`;

exports[`simple squash 1`] = `
Object {
  "commitsDetails": Array [
    Object {
      "message": "a

Some details",
      "sha": "0000000",
    },
    Object {
      "message": "b",
      "sha": "1111111",
    },
    Object {
      "message": "squash! a

Some more details",
      "sha": "2222222",
    },
  ],
  "expectedMessages": Array [
    "a

Some details

squash! a

Some more details",
    "b",
  ],
}
`;
